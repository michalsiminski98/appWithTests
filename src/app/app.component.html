<app-add-form></app-add-form>
<app-hide-button>SHOW / HIDE</app-hide-button>
<app-best-quote [title]= "titleForBest"></app-best-quote>
<app-quote-list [title] = "titleForList"></app-quote-list>
<app-worst-list [title] = "titleForWorst"></app-worst-list>