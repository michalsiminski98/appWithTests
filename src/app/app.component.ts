import { Component } from '@angular/core';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent {
  titleForBest: string = "Best Quote:"
  titleForList: string = "Quotes list:";
  titleForWorst: string = "Worst Quotes:";

  public isSubmitting: boolean = false;

  showForm = false;

  onShowForm(): void{
    this.showForm = !this.showForm;
  }

  adding(a:number, b:number): number{
    return a + b
  }
}

