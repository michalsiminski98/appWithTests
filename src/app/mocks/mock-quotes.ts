import { Quote } from "../interfaces/quote";

export const QUOTES: Quote[] = [
  {
    id: 0,
    author: "bardzo",
    quotation: "asddasdasads adsasd a dsads  a ddsa asd.",
    rate: -3,
  },
  {
    id: 1,
    author: "madry",
    quotation: "AEDRFDHSRHTSHDERTaSER EAWGAERWEWF WEEF.",
    rate: 3,
  },
  {
    id: 2,
    author: "autor",
    quotation: "039284120934 012 4012981284 0914 29810 928.",
    rate: 4,
  },
  {
    id: 3,
    author: "asgwfewefwd",
    quotation: "9832 jhfp934h4r2  309 r48h23  984rh3  q98 4rhq329r3h4r98 h23r908q2h3r98h98h.",
    rate: -5,
  },
]