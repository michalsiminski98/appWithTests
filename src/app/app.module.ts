import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';

import { AppComponent } from './app.component';
import { HideButtonComponent } from './hide-button/hide-button.component';
import { AddFormComponent } from './add-form/add-form.component';
import { QuoteListComponent } from './quote-list/quote-list.component';
import { BestQuoteComponent } from './best-quote/best-quote.component';
import { WorstListComponent } from './worst-list/worst-list.component';
import { FormsModule } from '@angular/forms';

@NgModule({
  declarations: [
    AppComponent,
    HideButtonComponent,
    AddFormComponent,
    QuoteListComponent,
    BestQuoteComponent,
    WorstListComponent
  ],
  imports: [
    BrowserModule,
    FormsModule
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
