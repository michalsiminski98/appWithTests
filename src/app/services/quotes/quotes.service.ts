import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { map } from 'rxjs/operators';
import { Quote } from '../../interfaces/quote';
import { QUOTES } from '../../mocks/mock-quotes';

@Injectable({
  providedIn: 'root',
})
export class QuotesService {
  private quotes: Quote[] = QUOTES.map((quote) => ({ ...quote }));

  public quotesSource = new BehaviorSubject<Quote[]>(this.quotes);

  public quotes$ = this.quotesSource.asObservable();
  public worstQuotes$ = this.quotes$.pipe(
    map((list) => list.filter((quote) => quote.rate < 0))
  );
  public bestQuote$ = this.quotes$.pipe(
    map((list) =>
      list.reduce((prev, curr) => {
        return prev.rate > curr.rate ? prev : curr;
      })
    )
  );

  constructor() {}

  // creating quote by form
  createQuote(author: string, quotation: string) {
    // creating new Quote with data from form
    const newQuote: Quote = {
      id: this.createQuoteId(),
      author,
      quotation,
      rate: 0,
    };
    // pushing new quote to service array
    this.quotes = [newQuote, ...this.quotes];
    this.quotesSource.next(this.quotes);
  }

  createQuoteId(): number {
    return Math.floor(Math.random() * 99999);
  }

  addingRate(id: number) {
    this.quotes = this.quotes.map((quote) => {
      if (quote.id === id) {
        return { ...quote, rate: quote.rate + 1 };
      }
      return quote;
    });
    this.quotesSource.next(this.quotes);
  }

  subtractingRate(id: number) {
    this.quotes = this.quotes.map((quote) => {
      if (quote.id === id) {
        return { ...quote, rate: quote.rate - 1 };
      }
      return quote;
    });
    this.quotesSource.next(this.quotes);
  }
}
