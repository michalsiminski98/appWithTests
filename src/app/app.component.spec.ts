import { of } from "rxjs";
import { Shallow } from "shallow-render";
import { AppComponent } from "./app.component";
import { AppModule } from "./app.module";
import { QUOTES } from "./mocks/mock-quotes";
import { QuotesService } from "./services/quotes/quotes.service";

describe('AppComponent', () => {
  let shallow: Shallow<AppComponent>;

  beforeEach(() => {
    shallow = new Shallow(
      AppComponent,
      AppModule
    ).provideMock([
      {
        provide: QuotesService,
        useValue: {
          quotes: jest.fn(() => of(QUOTES))
        }
      }
  ]);
  });

  it('should match snapshot', async () => {
    const { fixture } = await shallow.render();

    expect(fixture).toMatchSnapshot();
  })

  describe('Adding 2 values', () => {
    it('should return 4', async () => {
      const{inject, instance} = await shallow.render();
      expect(instance.adding(2,2)).toBe(4);
    })
  })

  describe('changing showForm status', () => {
    it('should be false by default', async () => {
      const{inject, instance} = await shallow.render();
      expect(instance.showForm).toBeFalsy();
    });
    it('should change "showForm" to true', async () => {
      const{inject, instance} = await shallow.render();
      instance.onShowForm();
      expect(instance.showForm).toBeTruthy();
    });
  })

});
