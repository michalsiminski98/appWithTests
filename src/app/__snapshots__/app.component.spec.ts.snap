// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should match snapshot 1`] = `
<ng-component>
  <app-root>
    <app-add-form />
    <app-hide-button>
      SHOW / HIDE
    </app-hide-button>
    <app-best-quote />
    <app-quote-list />
    <app-worst-list />
  </app-root>
</ng-component>
`;
